File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.py
import sys

n = int(sys.stdin.readline().rstrip())
numbers = list(map(int, sys.stdin.readline().rstrip().split(' ')))

lis = [0] * n
result = 0

for i in range(n):
    left = 0
    right = result

    while left < right:
        mid = (left + right) // 2

        if lis[mid] < numbers[i]:
            left = mid + 1
        else:
            right = mid

    lis[left] = numbers[i]

    if left == result:
        result += 1

print(result)

File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		int n = Integer.parseInt(br.readLine());
		int[] numbers =
			Arrays.stream(br.readLine().split(" "))
				.mapToInt(Integer::parseInt)
				.toArray();

		int[] lis = new int[n];
		int length = 0;

		for (int i = 0; i < n; i++) {
			int left = 0;
			int right = length;

			while (left < right) {
				int mid = (left + right) / 2;

				if (lis[mid] < numbers[i]) {
					left = mid + 1;
				} else {
					right = mid;
				}
			}

			lis[left] = numbers[i];

			if (left == length) {
				length++;
			}
		}

		System.out.println(length);
	}
}
